### PR TITLE
Fixed schema generation for subobjects

### PR DESCRIPTION
**Pull request by aleximprobable**
#### Description
Schema generator now properly generates schema files for subobjects. There's been an issue when schema generator worked with classes, that were derived from base classes, that had subobjects. As the schema files were invalid - the RPC didn't execute properly.
The link to the jira ticket: https://improbableio.atlassian.net/browse/UNR-1116

#### Release note
"Bugfix for schema generator: schema files should now be generated properly for subobjects of the Blueprint classes."

#### Tests
No tests added.

#### Documentation
Updated comments in code.

#### Primary reviewers
@improbable-valentyn @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 1]
```diff
@@ -277,8 +277,10 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
 				continue;
 			}
 
-			// Check whether the owner of this value is the CDO itself.
-			if (Value->GetOuter() == ContainerCDO)
+			// Check whether the owner is the CDO itself or the parent of the CDO.
+			// (it also covers cases when the CDO is a Blueprint derived from the owner class)
+			if(Value->GetOuter()->HasAnyFlags(RF_ClassDefaultObject) &&
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]:
```suggestion
			if (Value->GetOuter()->HasAnyFlags(RF_ClassDefaultObject) &&
```

Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: Make sure `Value->GetOuter() != nullptr` first.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 2]
```diff
@@ -277,8 +277,10 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
 				continue;
 			}
 
-			// Check whether the owner of this value is the CDO itself.
-			if (Value->GetOuter() == ContainerCDO)
+			// Check whether the owner is the CDO itself or the parent of the CDO.
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]:
```suggestion
			// Check whether the outer is the CDO itself or the parent of the CDO.
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 3]
```diff
@@ -277,8 +277,10 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
...
+			// (it also covers cases when the CDO is a Blueprint derived from the owner class)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]:
```suggestion
			// (it also covers cases when generating schema for a Blueprint derived from the outer class)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 4]
```diff
@@ -277,8 +277,13 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
...
 
-			// Check whether the owner of this value is the CDO itself.
-			if (Value->GetOuter() == ContainerCDO)
+			// Check whether the outer is the CDO of the class we're generating for
+			// or the CDO of any of its parent classes.
+			// (this also covers generating schema for a Blueprint derived from the outer's class)
+			UObject* Outer = Value->GetOuter();
+			if ((Outer != nullptr) &&
+				Outer->HasAnyFlags(RF_ClassDefaultObject) &&
+				ContainerCDO->GetClass()->IsChildOf(Outer->GetClass()))
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: Is this the same?
```suggestion
				ContainerCDO->IsA(Outer->GetClass())
```

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: These look similar and both produce the expected result. I will apply your change.

**Pull request merged**